### PR TITLE
Machine learning updates and refactor (duplicate of #54)

### PR DESCRIPTION
**Pull request by ma595**
This is a duplicate of #54.

Before merging check:

- [x] All tests pass. If tests do not pass, justify why and modify reference outputs. Check on CSD3 as well. 
  - [x] `test_init.py` 
  - [x] `test_test1.py`
  - [x] `test_test2.py`
  - [x] `test_task4.py`
  - [x] Check on another system (CSD3)
  - [x] Evaluation and visualisation steps work for `DEF_Trunk` 
  - [x] Evaluation and visualisation works for `DEF_CNP2`. 
- [x] Create new reference output. Put this on Github for now https://github.com/ma595/SPINacc-results
- [x] Update reference input dataset https://github.com/ma595/SPINacc-data
- Once this PR is merged this will allow the following to be merged. 
  - #81 
  - #82 

Other additions: 
- [x] ML.py refactor - split up the `MLmap_multidim` function with  `extrapolate` and `evaluate` functions. Cleaned up comments.
- [x] Added new `MLacc_results.csv` reference to `tests/data`. 
- [x] Updated all tests to new file format and checks - more robust to file paths. Enabled the test of netcdf restart file correspondence. 
- [x] Toggle between parallel and serial.
- [x] Set seeds in sklearn (kmeans) and imblearn (SMOTE) functions 
- [x] Revert implementation so `MLacc_results.csv` can be overwritten if performance degrades.  
- [x] Fixed bug when using multiprocessing - random seed not set thereby affecting reproducibility tests. 
- [x] Added `pytest.ini` file
- [x] Description of algorithm - add comments to code where necessary. 
- [x] Revert factor of 2 increase in varlist.json
- [x] Deleted `old_train.py` and `debug_train.py`
- [x] Path() used everywhere. 
- [x] Use additional arguments in config : parallel, model_out. Latter is a boolean that specifies whether to output the trained model to output directory. 
- [x] Cleaned up `test_task4.py` so check is done against the new `MLacc_results.csv` file.
- [x] PEP8 changes to module names.

The following changes are made:

### Data preprocessing
- Simplify readvar.py
- Add more training variables `[LWdown_std', 'PSurf_std', 'Qair_std', 'SWdown_std', 'Snowf_mean','Snowf_std']`
- Restore the time dimension of the dataset with a monthly frequency to increase the input (X) data size. 
- Add standard scaling to preprocess the data before ML training

### ML modelling
- Implemented multithread parallelization to train a ML model per target variable in parallelization
- Add new ML algorithm options: NN, XGBoost, RandomForest, MLP, Lasso, Stacking Ensemble
- `Shotgun` or stacking ensemble chooses the best performing model for a given variable using cross validation. 
- Test the evaluation performance and compare it with the previous ML model
- Implement multithread parallelization to train a ML model per target variable in parallelization

### Software health
- Add `CHANGES.md` which summarises all additions up to now and performance. 
- Add Docstrings to functions
- Remove `ML_biomass.py` `ML_som_litter.py`
- Add `pytest.ini`
- More detailed checks in initialisation, `test_init.py`. 

As an aside we also explored the use of a new NN approach: 

- Investigated use of a Neural Network:
  - NN is found to give inferior performance compared to tree based approaches
  - Attempted to Implement and train a NN for each ipool (thereby increasing the dataset); tune the hyperparameters using a validation dataset
  - This required, for each, ipool, collect all input datasets into a single dataset for NN training

Fixes https://github.com/CALIPSO-project/SPINacc/issues/53 https://github.com/CALIPSO-project/SPINacc/issues/57 #87 #90

### CODE REVIEW

**Comment by ma595** [comment]
A few updates:

- I've rebased changes from main and am working through the changes
- First observation after rebasing is that none of the tests currently pass:
  - `Tamp` was excluded from packdata (causing `test_task1.py` to fail.)
  - Was there any justification to do this @tztsai? 
  - All temperature statistics were incorrect because `Tair` was being overwritten after converting to celsius
  - I added `Tamp` and reordered the code in `Cluster.py` and now `test_init.py`, `test_task1.py` and `test_task2.py` pass. 
- `test_task4.py` still fails - which isn't a big surprise
  - Debugging strategy was to output the combine_XY between `main` and `nn` branches. Obviously differences emerge:
  - We conclude that this is because there are some additional variables:
  - `[LWdown_std', 'PSurf_std', 'Qair_std', 'SWdown_std', 'Snowf_mean','Snowf_std']` 21 columns vs 27 columns of data.
  - All other columns match between nn and main branches. 
  - The change is a simple one in `main.py`

```
-labx = ["Y"] + var_pred_name + ["pft"]
+labx = ["Y"] + list(packdata.data_vars) + ["pft"]
```

```
Index(['Unnamed: 0', 'GS_length', 'LAI0', 'LWdown_mean', 'LWdown_std', 'NPP0',
       'PSurf_mean', 'PSurf_std', 'Pre_GS', 'Qair_mean', 'Qair_std',
       'Rainf_mean', 'Rainf_std', 'SWdown_mean', 'SWdown_std', 'Snowf_mean',
       'Snowf_std', 'Tamp', 'Temp_GS', 'Tmax', 'Tmean', 'Tmin', 'Tstd', 'Y',
       'clay_frac', 'interx1', 'interx2'],
      dtype='object')
Index(['Unnamed: 0', 'GS_length', 'LAI0', 'LWdown_mean', 'NPP0', 'PSurf_mean',
       'Pre_GS', 'Qair_mean', 'Rainf_mean', 'Rainf_std', 'SWdown_mean', 'Tamp',
       'Temp_GS', 'Tmax', 'Tmean', 'Tmin', 'Tstd', 'Y', 'clay_frac', 'interx1',
       'interx2'],
      dtype='object')
```

**Comment by ma595** [comment]
Current output:
`som,1,2,0.9815856373172269,4.149280303938662,0.9931318635047343,4.390644416493333e-06,`

**Review by omarjamil — APPROVED** [review]
